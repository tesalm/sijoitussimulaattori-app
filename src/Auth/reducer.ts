import { User } from '../models';
import { ActionType, AuthAction } from './actions';

export interface Auth {
  user?: User;
  error?: Error;
}

const initialState: Auth = {};

export const authReducer = (
  state: Auth = initialState,
  action: AuthAction
): Auth => {
  switch (action.type) {
    case ActionType.LoginRequest:
      return state;
    case ActionType.LoginSuccess:
      return {
        ...state,
        user: action.user,
        error: undefined,
      };
    case ActionType.LoginFailure:
      return {
        ...state,
        user: undefined,
        error: action.error,
      };
    case ActionType.LogoutRequest:
      return initialState;
    default:
      return state;
  }
};
