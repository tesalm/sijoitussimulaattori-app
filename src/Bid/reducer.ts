import { ActionType, BidAction } from './actions';

export interface BidInfo {
  action: string;
  bidLevel: number;
  sumOfStocks: number;
  selectedPortfolio: string;
  symbol: string;
}

export const initialState: BidInfo = {
  action: '',
  bidLevel: 0,
  sumOfStocks: 0,
  selectedPortfolio: '',
  symbol: '',
};

export const bidReducer = (
  state: BidInfo = initialState,
  action: BidAction
): BidInfo => {
  switch (action.type) {
    case ActionType.UpdateBidInfo:
      return {
        ...state,
        action: action.actionType,
        bidLevel: action.bidLevel,
        sumOfStocks: action.sumOfStocks,
        selectedPortfolio: action.selectedPortfolio,
        symbol: action.symbol,
      };
    default:
      return state;
  }
};
