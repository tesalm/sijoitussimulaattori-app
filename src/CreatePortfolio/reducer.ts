import { ActionType, CreatePortfolioAction } from './actions';
import cloneDeep = require('lodash/cloneDeep');
import firebase, { auth } from 'react-native-firebase';

export interface CreatingPortfolio {
  loading: boolean;
  error?: Error;
  success: boolean;
}

const initialState: CreatingPortfolio = {
  loading: false,
  error: undefined,
  success: false,
};

export const createPortfolioReducer = (
  state: CreatingPortfolio = initialState,
  action: CreatePortfolioAction
): CreatingPortfolio => {
  switch (action.type) {
    case ActionType.CreatePortfolioBegin:
      return { ...state, loading: true, error: undefined, success: false };
    case ActionType.CreatePortfolioSuccess:
      return { ...state, loading: false, error: undefined, success: true };
    case ActionType.CreatePortfolioFailure:
      return { ...state, loading: false, error: action.error, success: false };
    default:
      return state;
  }
};
