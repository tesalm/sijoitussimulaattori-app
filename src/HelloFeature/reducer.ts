import { CounterAction, ActionType } from './actions';

export interface Counter {
  counterValue: number;
}

const initialState: Counter = {
  counterValue: 0,
};

export const counterReducer = (
  state: Counter = initialState,
  action: CounterAction
): Counter => {
  switch (action.type) {
    case ActionType.AddToCounter:
      return { ...state, counterValue: state.counterValue + action.amount };
    case ActionType.RemoveFromCounter:
      return { ...state, counterValue: state.counterValue - 1 };
    default:
      return state;
  }
};
