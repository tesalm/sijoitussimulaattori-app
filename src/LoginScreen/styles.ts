import { StyleSheet } from 'react-native';

const Colors = {
  background: '#004D40',
  text: '#FFFFFF',
};

export const loginScreenStyles = StyleSheet.create({
  background: {
    // <View>
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'space-evenly',
    alignItems: 'center',
    backgroundColor: Colors.background,
  },
  text: {
    color: Colors.text,
    fontFamily: 'Roboto',
    fontSize: 17,
    textAlign: 'center',
    marginBottom: '4%',
  },
  logo: {
    width: '60%',
    height: '30%',
    resizeMode: 'contain',
    marginTop: '8%',
  },
  buttonContainer: {
    // <View>
    width: '65%',
    marginBottom: '8%',
  },
});
