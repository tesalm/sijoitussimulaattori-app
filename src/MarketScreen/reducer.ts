import cloneDeep from 'lodash/cloneDeep';

import { ActionType, StockAction } from './actions';

export interface StockMetadata {
  symbol: string;
  name: string;
  type: string;
  region: string;
  marketOpen: string;
  marketClose: string;
  timeZone: string;
  currency: string;
  fetchTime: Date;
}

export interface DailyQuote {
  symbol: string;
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface Intraday {
  intradayQuote: DailyQuote[];
  fetchTime: Date;
}

export interface HistoryData {
  historyDataQuote: DailyQuote[];
  fetchTime: Date;
}

export interface SingleStock {
  stockMetadata?: StockMetadata;
  intraday?: Intraday;
  historyData?: HistoryData;
  metaLoading: boolean;
  intraLoading: boolean;
  historyLoading: boolean;
  refreshing: boolean;
  metaError?: Error;
  intraError?: Error;
  historyError?: Error;
}

export interface Stock {
  symbol: string;
  name: string;
  high: number;
  low: number;
  revenue: number;
  close: number;
  stockInfo?: SingleStock;
}

export interface StocksListing {
  stocks: Array<Stock>;
  loading: boolean;
  refreshing: boolean;
  error?: Error;
  currentSymbol?: string;
}

const defaultSingleStock: SingleStock = {
  metaLoading: false,
  intraLoading: false,
  historyLoading: false,
  refreshing: false,
};

export const initialState: StocksListing = {
  stocks: [],
  loading: false,
  refreshing: false,
  error: undefined,
  currentSymbol: undefined,
};

export const stocksListingReducer = (
  state: StocksListing = initialState,
  action: StockAction
): StocksListing => {
  switch (action.type) {
    case ActionType.RequestStocksBegin:
      return {
        ...cloneDeep(state),
        loading: true,
        refreshing: true,
        error: undefined,
      };
    case ActionType.RequestStocksSuccess:
      return {
        ...state,
        stocks: action.stocks,
        loading: false,
        refreshing: false,
        error: undefined,
      };
    case ActionType.RequestStocksFailure:
      return {
        ...cloneDeep(state),
        loading: false,
        refreshing: false,
        error: action.error,
      };

    case ActionType.SaveAsCurrentSymbol:
      return { ...cloneDeep(state), currentSymbol: action.symbol };

    case ActionType.GetStockMetadataBegin: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        metaLoading: true,
      };
      return { ...state, stocks: stockList };
    }
    case ActionType.GetStockMetadataSuccess: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        stockMetadata: { ...action.metaData, fetchTime: action.fetchTime },
        metaLoading: false,
        metaError: undefined,
      };
      return { ...state, stocks: stockList };
    }
    case ActionType.GetStockMetadataFailure: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        stockMetadata: undefined,
        metaLoading: false,
        metaError: action.error,
      };
      return { ...state, stocks: stockList };
    }

    case ActionType.GetIntradayBegin: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        intraLoading: true,
      };
      return { ...state, stocks: stockList };
    }
    case ActionType.GetIntradaySuccess: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        intraday: {
          intradayQuote: action.intraData,
          fetchTime: action.fetchTime,
        },
        intraLoading: false,
        intraError: undefined,
      };
      return { ...state, stocks: stockList };
    }
    case ActionType.GetIntradayFailure: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        intraday: undefined,
        intraLoading: false,
        intraError: action.error,
      };

      return { ...state, stocks: stockList };
    }

    case ActionType.RefreshIntradayBegin: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        refreshing: true,
      };
      return { ...state, stocks: stockList };
    }
    case ActionType.RefreshIntradaySuccess: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        intraday: {
          intradayQuote: action.intraData,
          fetchTime: action.fetchTime,
        },
        refreshing: false,
        intraError: undefined,
      };

      return { ...state, stocks: stockList };
    }
    case ActionType.RefreshIntradayFailure: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        intraday: undefined,
        refreshing: false,
        intraError: action.error,
      };
      return { ...state, stocks: stockList };
    }

    case ActionType.GetHistoryBegin: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        historyLoading: true,
      };
      return { ...state, stocks: stockList };
    }
    case ActionType.GetHistorySuccess: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        historyData: {
          historyDataQuote: action.historyData,
          fetchTime: action.fetchTime,
        },
        historyLoading: false,
        historyError: undefined,
      };
      return { ...state, stocks: stockList };
    }
    case ActionType.GetHistoryFailure: {
      const stockList = cloneDeep(state.stocks);
      const stockIndex = getStockIndex(stockList, action.symbol);
      if (stockIndex < 0) {
        return { ...state };
      }

      stockList[stockIndex].stockInfo = {
        ...(stockList[stockIndex].stockInfo || defaultSingleStock),
        historyData: undefined,
        historyLoading: false,
        historyError: action.error,
      };
      return { ...state, stocks: stockList };
    }

    default:
      return state;
  }
};

const getStockIndex = (stockList: Stock[], symbol: string) =>
  stockList.findIndex((stock) => {
    return stock.symbol === symbol;
  });
