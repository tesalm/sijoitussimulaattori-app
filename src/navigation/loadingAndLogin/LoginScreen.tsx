import React from 'react';
import { NavigationScreenProps } from 'react-navigation';

import AnonymousLoginField from './AnonymousLoginField';

export class LoginScreen extends React.Component<NavigationScreenProps> {
  constructor(props: NavigationScreenProps){
    super(props);
  }

  render(){ return <AnonymousLoginField />; }
}
