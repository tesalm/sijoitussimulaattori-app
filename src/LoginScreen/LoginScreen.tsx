import React from 'react';
import { Image, ImageStyle, Text, View } from 'react-native';
import { NavigationScreenProps } from 'react-navigation';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { t } from '../assets/i18n';
import { login } from '../Auth/actions';
import { LoginState } from '../Auth/reducer';
import { Button } from '../input/button';
import { RootState } from '../redux/reducers';
import { ModalActivityIndicator } from './components/ModalActivityIndicator';
import { loginScreenStyles } from './styles';

export interface LoginViewProps {
  loginState: LoginState;
  loginError?: Error;
  onLoginAsAnonym: typeof login;
}

type LoginViewPropsWithNavigation = LoginViewProps & NavigationScreenProps;

class LoginScreen extends React.Component<LoginViewPropsWithNavigation> {
  constructor(props: LoginViewPropsWithNavigation) {
    super(props);
  }

  render() {
    const { onLoginAsAnonym, loginError, loginState } = this.props;

    const loginButtonText = loginError
      ? t('LoginScreen.AfterErrButton')
      : t('LoginScreen.NewUserButton');
    const loginGreeting = loginError
      ? loginError.message
      : t('LoginScreen.NewUserGreeting');

    return (
      <View style={loginScreenStyles.background}>
        {// Show modal activity indicator on top of everything else when logging in.
        loginState === LoginState.LoggingIn && <ModalActivityIndicator />}
        {/*Logo.*/}
        <Image
          style={loginScreenStyles.logo as ImageStyle}
          source={require('../assets/images/Logo.png')}
        />
        {/*New user greetee and anonymous login button.*/}
        <View style={loginScreenStyles.buttonContainer}>
          <Text style={loginScreenStyles.text}>{loginGreeting}</Text>
          <Button
            buttonText={loginButtonText}
            onPress={onLoginAsAnonym}
            lightBackground={true}
          />
        </View>
      </View>
    );
  }
}

const mapStateToProps = (state: RootState) => ({
  loginState: state.login.loginState,
  loginError: state.login.loginError,
});
const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      onLoginAsAnonym: login,
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginScreen);

export { LoginScreen as LoginScreenTest };
