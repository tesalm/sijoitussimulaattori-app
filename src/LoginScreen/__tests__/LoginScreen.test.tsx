import * as React from 'react';
import renderer from 'react-test-renderer';

import { LoginState } from '../../Auth/reducer';
import { LoginScreenTest, LoginViewProps } from '../LoginScreen';

describe('Testing Login Screen', () => {
  //Mock for navigation.
  const navigationMock: any = {};

  const defaultLoginViewProps: LoginViewProps = {
    loginState: LoginState.LoggedOut,
    loginError: undefined,
    onLoginAsAnonym: jest.fn()
  }

  const afterLoginErrorProps: LoginViewProps = {
    loginState: LoginState.LoggedOut,
    loginError: Error('fake-failure occurred'),
    onLoginAsAnonym: jest.fn()
  }

  const duringLoggingInProps: LoginViewProps = {
    loginState: LoginState.LoggingIn,
    loginError: undefined,
    onLoginAsAnonym: jest.fn()
  }

  it('renders correctly', async () => {
    const component = renderer
      .create(<LoginScreenTest {...defaultLoginViewProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly after login error', async () => {
    const component = renderer
      .create(<LoginScreenTest {...afterLoginErrorProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly when logging in and waiting', async () => {
    const component = renderer
      .create(<LoginScreenTest {...duringLoggingInProps} {...navigationMock} />)
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});