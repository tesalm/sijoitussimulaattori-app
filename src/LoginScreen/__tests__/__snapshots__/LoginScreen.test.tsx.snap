// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login Screen renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#004D40",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-evenly",
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../resources/Logo.png",
      }
    }
    style={
      Object {
        "height": "30%",
        "marginTop": "8%",
        "resizeMode": "contain",
        "width": "60%",
      }
    }
  />
  <View
    style={
      Object {
        "marginBottom": "8%",
        "width": "65%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "fontSize": 17,
          "marginBottom": "4%",
          "textAlign": "center",
        }
      }
    >
      Your first time at the ValueTown?
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 5,
            "elevation": 3,
            "fontSize": 16,
            "justifyContent": "center",
            "padding": 16,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#004D40",
              "fontFamily": "Roboto",
            }
          }
        >
          CONTINUE AS NEW USER
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Testing Login Screen renders correctly after login error 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#004D40",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-evenly",
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../resources/Logo.png",
      }
    }
    style={
      Object {
        "height": "30%",
        "marginTop": "8%",
        "resizeMode": "contain",
        "width": "60%",
      }
    }
  />
  <View
    style={
      Object {
        "marginBottom": "8%",
        "width": "65%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "fontSize": 17,
          "marginBottom": "4%",
          "textAlign": "center",
        }
      }
    >
      fake-failure occurred
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 5,
            "elevation": 3,
            "fontSize": 16,
            "justifyContent": "center",
            "padding": 16,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#004D40",
              "fontFamily": "Roboto",
            }
          }
        >
          TRY AGAIN
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Testing Login Screen renders correctly when logging in and waiting 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#004D40",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-evenly",
    }
  }
>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={true}
  >
    <View
      style={
        Object {
          "backgroundColor": "#004D40E9",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#FFFFFF"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </Modal>
  <Image
    source={
      Object {
        "testUri": "../../../resources/Logo.png",
      }
    }
    style={
      Object {
        "height": "30%",
        "marginTop": "8%",
        "resizeMode": "contain",
        "width": "60%",
      }
    }
  />
  <View
    style={
      Object {
        "marginBottom": "8%",
        "width": "65%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "fontSize": 17,
          "marginBottom": "4%",
          "textAlign": "center",
        }
      }
    >
      Your first time at the ValueTown?
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 5,
            "elevation": 3,
            "fontSize": 16,
            "justifyContent": "center",
            "padding": 16,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#004D40",
              "fontFamily": "Roboto",
            }
          }
        >
          CONTINUE AS NEW USER
        </Text>
      </View>
    </View>
  </View>
</View>
`;
